### ISSUE TITLE
`dataset_test` picks up time values as unallowed

### ISSUE THREAD

**Issue by yangsophieee** [traitecoevo/traits.build#87]
Currently `dataset_test` fails when substitutions for time variables (e.g. `flowering_time`, `fruiting_time`) exist, because of [this line](https://github.com/traitecoevo/traits.build/blob/53d56ef18051e38808ecf6ee7e3e261e3f38307e/R/testdata.R#L666) looking for whether values are found within the trait definitions. 

I can think of several easy ways to fix this: if the values only contain "n" and "y" then you don't check for the trait definitions. 

This made me think about [this line](https://github.com/traitecoevo/traits.build/blob/53d56ef18051e38808ecf6ee7e3e261e3f38307e/R/process.R#L1024) in `process.R` which specifically codes for the traits `flowering_time` and `fruiting_time`. But this isn't very generalisable to other databases if they want to make their own time traits. I wonder if it's worth making a new "time" category of traits, in addition to "categorical" and "numeric". What do you think @ehwenk?

**Comment by ehwenk** [comment]
I've previously thought about having a separate trait category for `flowering_time`, etc. for similar reasons. Because when I've added a new time-trait (`recruitment_time`), you're right, I then had to tweak code (I think for building the dictionary ontology). And it probably should have been added to this list as well. 

The specific test you referenced searches for `N`'s and `Y`'s only, which is specific to time. 

But a separate problem with the `time` traits is that they are categorical traits for which the allowable trait values aren't listed. That could be true to non-time traits - given the trait `recruitment_time` isn't causing any errors, it seems we've already adapted the traits.build workflow to handle instances where categorical traits don't have set values. Maybe this is actually only a problem for the code to build the APD, in which case I guess it is fine to require a fix.

@dfalster what do you think? Is it a `time` category we should consider adding or a `factor` (or other word) category for traits without set values. Or nothing?

**Comment by dfalster** [comment]
Backlog triage: **the stopgap landed, the actual design question is still open, and there is a concrete piece of AusTraits-specific code left in the generic package.**

**What was done:** the narrow fix the issue itself suggested. `R/testdata.R:653` now excludes 12-character Y/N strings from categorical-definition checking:

```r
to_check <- to_check[!(grepl("^[YyNn]+$", to_check) & stringr::str_length(to_check) == 12)]
```

**What wasn't:** there is still no `time` trait type — `grep '"time"' R/*.R` returns zero hits. Trait `type` is branched on as exactly two values throughout (`R/process.R:1134`, `:1161`, `:1197`, `:1272-1273`, `R/testdata.R:653`).

And the specific line that prompted this issue is unchanged. `R/process.R:1145`:

```r
if (trait %in% c("flowering_time", "fruiting_time")) {
```

which then applies "Time can only contain Y & Ns" (`:1152`) and "Times must be length 12" (`:1157`). So two AusTraits trait names are hardcoded inside the genericised package. Consequences worth noting:

- a downstream database defining its own time trait gets **no** validation of these values
- `recruitment_time` (which @ehwenk raised above) is silently unvalidated even within AusTraits
- a database that happens to define `flowering_time` differently gets AusTraits' rules imposed on it

**Blocking decision** — @ehwenk's question to @dfalster above was never answered:

> is it best to add another category of trait, `time`, or to have a subcategory of `categorical` traits that are `factors` [with] no specific list of `allowed values`?

Those imply quite different amounts of work. Worth resolving, because two things follow regardless of which way it goes: de-hardcoding `R/process.R:1145` so the rule is driven by trait metadata rather than by name, and confirming whether APD-building is the only remaining breakage.

Related: #49 / #220 — same theme one layer down, where `read_csv` reformats time strings and the metadata recorded seconds instead. Substitutions on a pure-time column are still affected by that (`R/process.R:1497`, `:1552` coerce an `hms` column with `as.character()`, so a metadata `find: 7:00:00` never matches `"07:00:00"`), which is the case David works around with `custom_R_code`. That may be worth folding into whatever is decided here.

**Comment by ehwenk** [comment]
I do not think this is actually about a "time" category at this point. We don't need that. What might be good is to explicitly have an indication that a categorial trait does not have a list of defined allowed values. So far these are flowering time, fruiting time, recruitment time, but other instances could emerge and these all end up being hard-coded in as exceptions. 

There could be an extra field within the dictionary that is optional? That specifies a trait as a categorical trait without defined levels? That seems that most generalisable - and indeed might be desirable for some traits.build users who want open-ended trait values - e.g. taxon names as trait values.
